src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,776 @@
+        /// Kind of argument.
+        /// </summary>
+        ArgumentKind Kind { get; }
+        /// <summary>
+        /// Mode of argument.
+        /// </summary>
+        ArgumentMode Mode { get; }
+        /// <summary>
+        /// Value supplied for the argument.
+        /// </summary>
+        IExpression Value { get; }
+        /// <summary>
+        /// Conversion applied to the argument value passing it into the target method. Applicable only to In or Reference arguments.
+        /// </summary>
+        IExpression InConversion { get; }
+        /// <summary>
+        /// Conversion applied to the argument value after the invocation. Applicable only to Out or Reference arguments.
+        /// </summary>
+        IExpression OutConversion { get; }
+    }
+
+    /// <summary>
+    /// Kinds of arguments.
+    /// </summary>
+    public enum ArgumentKind
+    {
+        /// <summary>
+        /// Argument is specified positionally and matches the parameter of the same ordinality.
+        /// </summary>
+        Positional,
+        /// <summary>
+        /// Argument is specified by name and matches the parameter of the same name.
+        /// </summary>
+        Named,
+        /// <summary>
+        /// Argument becomes an element of an array that matches a trailing C# params or VB ParamArray parameter.
+        /// </summary>
+        ParamArray
+    }
+
+    /// <summary>
+    /// Modes of arguments.
+    /// </summary>
+    public enum ArgumentMode
+    {
+        /// <summary>
+        /// Argument is passed as an input value only.
+        /// </summary>
+        In,
+        /// <summary>
+        /// Argument is passed without an input value and is assigned a value by the time the invoked method returns.
+        /// </summary>
+        Out,
+        /// <summary>
+        /// Argument is passed by reference.
+        /// </summary>
+        Reference
+    }
+
+    /// <summary>
+    /// Represents a named argument.
+    /// </summary>
+    public interface INamedArgument : IArgument
+    {
+        /// <summary>
+        /// Name of the parameter that the argument matches.
+        /// </summary>
+        string Name { get; }
+    }
+
+    /// <summary>
+    /// Represents a reference, which refers to a symbol or an element of a collection.
+    /// </summary>
+    public interface IReference : IExpression
+    {
+        /// <summary>
+        /// Kind of the reference.
+        /// </summary>
+        ReferenceKind ReferenceKind { get; }
+    }
+
+    /// <summary>
+    /// Kinds of references.
+    /// </summary>
+    public enum ReferenceKind
+    {
+        Local,
+        Parameter,
+        Temporary,
+        ArrayElement,
+        PointerIndirection,
+        StaticField,
+        InstanceField,
+        StaticMethod,
+        InstanceMethod,
+        StaticPropertyGet,
+        InstancePropertyGet,
+        StaticPropertySet,
+        InstancePropertySet,
+        StaticProperty,
+        InstanceProperty,
```

Comment:
Static or Instance can be determined from the associated Field, Method, or Property.